utils/zerovec/src/varzerovec/mod.rs

```diff
@@ -0,0 +1,417 @@
+    /// assert_eq!(iter_results[0], "foo");
+    /// assert_eq!(iter_results[1], "bar");
+    /// assert_eq!(iter_results[2], "baz");
+    /// assert_eq!(iter_results[3], "quux");
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn iter<'b: 'a>(&'b self) -> impl Iterator<Item = &'b T::VarULE> {
+        // We use Either here so that we can use `impl Trait` with heterogeneous types
+        //
+        // An alternate design is to write explicit iterators. Once Rust has generators this will
+        // be unnecessary.
+        match self.0 {
+            VarZeroVecInner::Owned(ref vec) => Either::Left(vec.iter().map(|t| t.as_unaligned())),
+            VarZeroVecInner::Borrowed(components) => Either::Right(components.iter()),
+        }
+    }
+
+    /// Get one of VarZeroVec's elements, returning None if the index is out of bounds
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::VarZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo".to_owned(), "bar".to_owned(),
+    ///                    "baz".to_owned(), "quux".to_owned()];
+    /// let bytes = VarZeroVec::get_serializable_bytes(&strings);
+    /// let mut vec: VarZeroVec<String> = VarZeroVec::try_from_bytes(&bytes)?;
+    ///
+    /// let mut iter_results: Vec<&str> = vec.iter().collect();
+    /// assert_eq!(vec.get(0), Some("foo"));
+    /// assert_eq!(vec.get(1), Some("bar"));
+    /// assert_eq!(vec.get(2), Some("baz"));
+    /// assert_eq!(vec.get(3), Some("quux"));
+    /// assert_eq!(vec.get(4), None);
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn get<'b>(&'b self, idx: usize) -> Option<&'b T::VarULE> {
+        match self.0 {
+            VarZeroVecInner::Owned(ref vec) => vec.get(idx).map(|t| t.as_unaligned()),
+            VarZeroVecInner::Borrowed(components) => components.get(idx),
+        }
+    }
+
+    /// Convert this into a mutable vector of the owned `T` type, cloning if necessary.
+    ///
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::str::Utf8Error;
+    /// # use zerovec::VarZeroVecError;
+    /// # use zerovec::VarZeroVec;
+    ///
+    /// let strings = vec!["foo".to_owned(), "bar".to_owned(),
+    ///                    "baz".to_owned(), "quux".to_owned()];
+    /// let bytes = VarZeroVec::get_serializable_bytes(&strings);
+    /// let mut vec: VarZeroVec<String> = VarZeroVec::try_from_bytes(&bytes)?;
+    ///
+    /// assert_eq!(vec.len(), 4);
+    /// let mutvec = vec.make_mut();
+    /// mutvec.push("lorem ipsum".into());
+    /// mutvec[2] = "dolor sit".into();
+    /// assert_eq!(&vec[0], "foo");
+    /// assert_eq!(&vec[1], "bar");
+    /// assert_eq!(&vec[2], "dolor sit");
+    /// assert_eq!(&vec[3], "quux");
+    /// assert_eq!(&vec[4], "lorem ipsum");
+    /// # Ok::<(), VarZeroVecError<Utf8Error>>(())
+    /// ```
+    pub fn make_mut(&mut self) -> &mut Vec<T>
```

Comment:
I don't see the use case for this function.  I figure that you would call `.to_vec()` to get your `Vec<T>` and then `.from_aligned()` to convert that back to a `VarZeroVec<T>` when done.

Reply:
See `Arc::make_mut()` and `Cow::to_mut()`, this basically gives us convenient copy-on-write semantics.

In our case this is great for patching, because you can call `make_mut()` and then do your patching operation on the resultant vector.